File: queue/queue.cxx
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <queue.h>


// ----------------------------------------------------------------------------
Queue * create_queue(int size)
{
    int * array = (int *) malloc(sizeof(int) * size);
    assert(array != NULL);

    Queue *queue = new Queue;
    queue->array = array;
    queue->head = 0;
    queue->tail = 0;
    queue->max_size = size;
    queue->current_size = 0;

    return queue;
}

// ----------------------------------------------------------------------------
void clear_queue(Queue *queue)
{
    free(queue->array);
}

// ----------------------------------------------------------------------------
bool is_full(Queue *queue)
{
    if (queue->max_size == queue->current_size)
        return true;
    else
        return false;
}

// ----------------------------------------------------------------------------
bool is_empty(Queue *queue)
{
    if(queue->current_size == 0)
        return true;
    else
        return false;
}

// ----------------------------------------------------------------------------
int enqueue(Queue *queue, int data)
{
    if (!is_full(queue))
    {
        queue->array[queue->tail] = data;
        queue->tail++;
        queue->current_size++;

        if (queue->tail == queue->max_size)
        {
            queue->tail = 0;
        }
        return 1;
    }
    else
    {
        printf("Error - the queue if full \n");
        return NULL;
    }
}

// ----------------------------------------------------------------------------
int dequeue(Queue *queue)
{
    if(!is_empty(queue))
    {
        int data = queue->array[queue->head];
        queue->head++;
        queue->current_size--;

        if (queue->head == queue->max_size)
        {
            queue->head = 0;
        }
        return data;
    }
    else
    {
        printf("Error - the queue is empty \n");
        return NULL;
    }
}

// ----------------------------------------------------------------------------
void print_queue(Queue *queue)
{
    printf("Queue => [");
    for (int i = 0; i < queue->max_size - 1; i++)
    {
        printf("%d, ", *(queue->array + i));
    }
    printf("%d]", *(queue->array + queue->max_size - 1));
    printf("\n");
}

File: queue/queue.py
class Queue(object):
    """
    Implements a Circular Queue data structure
    """

    def __init__(self, size):
        self.max_size = size
        self.array = [None] * self.max_size
        self.head = 0
        self.tail = 0
        self.current_size = 0

    def __str__(self):
        s = str([x for x in self.array])
        return "Queue => " + s

    def is_full(self):
        """
        Test if the Queue is full.
        """
        if self.current_size == self.max_size:
            return True
        else:
            return False

    def is_empty(self):
        """
        Test if the Queue is empty.
        """
        if self.current_size == 0:
            return True
        else:
            return False

    def enqueue(self, data):
        """
        Inserts the specified element into this queue, if the capacity allows this
        operation. By definition the data is appended to the queue.
        """
        if not self.is_full():
            self.array[self.tail] = data
            self.tail += 1
            self.current_size += 1

            if self.tail == self.max_size:
                self.tail = 0
        else:
            print("Error - The Queue is full")
            return False

    def dequeue(self):
        """
        Retrieves the head of this Queue.
        """
        if not self.is_empty():
            data = self.array[self.head]
            self.head += 1
            self.current_size -= 1

            if self.head == self.max_size:
                self.head = 0

            return data
        else:
            print("Error - The Queue is empty")
            return False
